### PR TITLE
feat(channel): addition of a chunked channel

### PR DESCRIPTION
**Pull request by pyrho**
A new channel implementation has been created to allow sending &
receiving messages in chunks for underlying channel implementations that
cannot sustain large messages.

BREAKING CHANGES:
* GenericChannel now has a dedicate constructor which accepts an
optional timeout argumet that defaults to 5000ms
* GenericChannel's `_timeout` is now private, and a `timeout` getter is
available

### CODE REVIEW

**Comment by pyrho** [comment]
For the record this was done to workaround an issue in Edge44 where sending "large" messages through the Runtime.Port instance would break everything.

See the Edge bug report [here](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/19445073/) .

File: src/Transport.ts [hunk 1]
```diff
@@ -172,7 +170,7 @@ export class Transport {
      * and rejection function
      *
      */
-    private _registerRemoteHandler({ slotName }: TransportMessage): void {
+    private _registerRemoteHandler({ slotName }: any): void {
```
Comment (highfivedenis) [inline src/Transport.ts]: Could you restore the typing?

**Comment by highfivedenis** [comment]
👏 for the PR.

Nevertheless, I have doubts that the `ChunkedChannel.ts` should be in the main lib, even if it's light there is no need to add it.

I would rather prefer to see it in the example of Channel `https://github.com/Dashlane/ts-event-bus/tree/master/examples/channels`. Saying that made me realize that we should also export example channel well-testing like yours, but IMO not in the main lib.

**Comment by wdelmas** [comment]
@highfivedenis I think that `ChunkedChannel` should be the default channel and be renamed `GenericChannel` since the actual Generic channel is a `ChunkedChannel`  but with only one chunk^^

**Comment by pyrho** [comment]
@wdelmas 
> I think that `ChunkedChannel` should be the default channel and be renamed `GenericChannel` since the actual Generic channel is a `ChunkedChannel` but with only one chunk^^

That makes sense, and the way it's currently implemented if the payload size is smaller than the required `chunkSize`, the `ChunkedChannel` has essentially the behavior as the `GenericChannel`.

@highfivedenis 
> Nevertheless, I have doubts that the ChunkedChannel.ts should be in the main lib, even if it's light there is no need to add it.

I agree and was reluctant at first to include this as part of the main lib, because it's tied to a specific need.
I don't think adding it as an example would help since they aren't exported/readily usable by consumers; clients using the `ChunkedChannel` would need to reimplement it (on both sides of the channel).
To solve this, we could have a separate package (kinda like a "plugin") that would export the chunked channel, but that seems overkill and cumbersome to use and maintain.

In the end I chose to include it as part of the lib because it's a utility that is useful to consumers of tsEventBus, if they choose to. And the added size to the package is negligible.

What would you suggest instead of making this an example rather than an exported class? Would @wdelmas 's suggestion be acceptable?

File: src/Channels/ChunkedChannel.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
+import { GenericChannel } from './GenericChannel'
+import { TransportMessage } from '../Message'
+
+export type ChunkedMessageStart = { type: 'chunk_start', chunkId: string, size: number }
+export type ChunkedMessage = { type: 'chunk_data', chunkId: string, data: any }
+export type ChunkedMessageEnd = { type: 'chunk_end', chunkId: string }
```
[the remark this answers is not in the document]
Comment (pyrho) [reply src/Channels/ChunkedChannel.ts]: I initially put them there, and moved them to this file in the end.
Types in `Message.ts` are all purely transport related. These types on the other hand are solely consumed by the `ChunkedChannel`, in other words, they are internal (to the `ChunkedChannel`) transport messages.

File: src/Channels/ChunkedChannel.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        sendChunks()
+
+        this._sender({
+            type: 'chunk_end',
+            chunkId
+        })
+
+    }
+
+    protected _messageReceived(message: TransportMessage | ChunkedTransportMessage) {
```
[the remark this answers is not in the document]
Comment (pyrho) [reply src/Channels/ChunkedChannel.ts]: This implementation of `_messageReceived`  literally processes either a `TransportMessage` if the message doesn't need chunking, or a `ChunkedTransportMessage` if the message has been chunked.

I think the verbosity helps with undestanding in this case.

File: src/Channels/ChunkedChannel.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    private _receiveNewChunk(message: ChunkedMessageStart) {
+        this._buffer[message.chunkId] = {
+            id: message.chunkId,
+            chunks: [],
+            size: message.size
+        }
+    }
+
+    private _receiveChunkData(message: ChunkedMessage) {
+        this._buffer[message.chunkId].chunks.push(message.data)
```
[the remark this answers is not in the document]
Comment (pyrho) [reply src/Channels/ChunkedChannel.ts]: Indeed, I will add checks here.

File: src/Transport.ts (hunk 1 above)
Comment (pyrho) [reply src/Transport.ts]: 👌

**Comment by highfivedenis** [comment]
@pyrho I would rather prefer to keep the simplistic `GenericChannel`. I will create an issue in order to split `Channels` from the main bundle. Thanks for the PR.

File: src/Channels/ChunkedChannel.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (wdelmas) [reply src/Channels/ChunkedChannel.ts]: make sense

[the remark this answers is not in the document]
Comment (pyrho) [reply src/Channels/ChunkedChannel.ts]: We could do that indeed, but I find that it's easier to read/understand the way it is right now.

**Pull request merged**